crates/node_binding/src/resolver_factory.rs

```diff
@@ -26,48 +27,66 @@ impl JsResolverFactory {
     })
   }
 
-  pub fn get_resolver_factory(&mut self, resolve_options: Resolve) -> Arc<ResolverFactory> {
+  pub fn get_resolver_factory(
+    &mut self,
+    resolve_options: Resolve,
+    pnp_manifest_path: Option<PathBuf>,
+  ) -> Arc<ResolverFactory> {
     match &self.resolver_factory {
       Some(resolver_factory) => resolver_factory.clone(),
       None => {
-        let resolver_factory = Arc::new(ResolverFactory::new(resolve_options));
+        let pnp_manifest = pnp_manifest_path
+          .and_then(|x| load_pnp_manifest(x).ok())
+          .map(Arc::new);
+        let resolver_factory = Arc::new(ResolverFactory::new(resolve_options, pnp_manifest));
         self.resolver_factory = Some(resolver_factory.clone());
         resolver_factory
       }
     }
   }
 
-  pub fn get_loader_resolver_factory(&mut self, resolve_options: Resolve) -> Arc<ResolverFactory> {
+  pub fn get_loader_resolver_factory(
+    &mut self,
+    resolve_options: Resolve,
+    pnp_manifest_path: Option<PathBuf>,
+  ) -> Arc<ResolverFactory> {
     match &self.loader_resolver_factory {
       Some(resolver_factory) => resolver_factory.clone(),
       None => {
-        let resolver_factory = Arc::new(ResolverFactory::new(resolve_options));
+        let pnp_manifest = pnp_manifest_path
+          .and_then(|x| load_pnp_manifest(x).ok())
+          .map(Arc::new);
+        let resolver_factory = Arc::new(ResolverFactory::new(resolve_options, pnp_manifest));
         self.loader_resolver_factory = Some(resolver_factory.clone());
         resolver_factory
       }
     }
   }
 
-  #[napi(ts_args_type = "type: string, options?: RawResolveOptionsWithDependencyType")]
+  #[napi(
+    ts_args_type = "type: string, options?: RawResolveOptionsWithDependencyType,pnpManifestPath: string|undefined"
+  )]
   pub fn get(
     &mut self,
     r#type: String,
     raw: Option<RawResolveOptionsWithDependencyType>,
+    context: String,
   ) -> napi::Result<JsResolver> {
+    let pnp_manifest_path = find_closest_pnp_manifest_path(context);
```

Comment:
This API seems to have a recursive I/O operation. Is it possible to make it opt-in?

Reply:
then we may need feature flag in rspack like
```js
module.exports = {
   pnp: true | false
}
``

Reply:
actually this can be passed from js side by `require('module').findPnpApi` and enhanced-resolve call it for every resolve call(which is time consuming, https://github.com/webpack/enhanced-resolve/blob/247edebc909040ff99200108f7165e57f4fac2b1/lib/ResolverFactory.js#L138),but better performance for monorepo(like multi .pnp.cjs support)
I think we can support bultin resolve plugin so uers don't use pnp shouldn't pay the cost of pnpapi call
```js
module.exports = {
   resolve: {
      plugins: ['builtin:resolve-pnp-pugin']
   }
}
```

Reply:
Note that we could make the implementation smarter, and only check directories that haven't been checked before, assuming we keep a cache of some sort (a `HashMap<Path, Option<Path>>` would be enough); this would trim most of the recursion.

Reply:
do you think the cache logic should be implemented in `find_closest_pnp_manifest_path` or outside